clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,260 @@
+    }
+
+    // Visible for testing
+    HeartbeatRequestManager(
+        final Time time,
+        final LogContext logContext,
+        final ConsumerConfig config,
+        final CoordinatorRequestManager coordinatorRequestManager,
+        final SubscriptionState subscriptions,
+        final MembershipManager membershipManager,
+        final HeartbeatRequestState heartbeatRequestState,
+        final ErrorEventHandler nonRetriableErrorHandler) {
+        this.time = time;
+        this.logger = logContext.logger(this.getClass());
+        this.subscriptions = subscriptions;
+        this.rebalanceTimeoutMs = config.getInt(CommonClientConfigs.MAX_POLL_INTERVAL_MS_CONFIG);
+        this.coordinatorRequestManager = coordinatorRequestManager;
+        this.heartbeatRequestState = heartbeatRequestState;
+        this.membershipManager = membershipManager;
+        this.nonRetriableErrorHandler = nonRetriableErrorHandler;
+    }
+
+    @Override
+    public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
+        if (!coordinatorRequestManager.coordinator().isPresent() || notInGroup()) {
+            return new NetworkClientDelegate.PollResult(
+                Long.MAX_VALUE, Collections.emptyList());
+
+        }
+
+        if (!heartbeatRequestState.canSendRequest(currentTimeMs)) {
+            return new NetworkClientDelegate.PollResult(
+                heartbeatRequestState.nextHeartbeatMs(currentTimeMs),
+                Collections.emptyList());
+        }
+        this.heartbeatRequestState.onSendAttempt(currentTimeMs);
+        NetworkClientDelegate.UnsentRequest request = makeHeartbeatRequest();
+        // return Long.MAX_VALUE because we will update the timer when the response is received
+        return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.singletonList(request));
+    }
+
+    boolean notInGroup() {
+        return membershipManager.state() == MemberState.UNJOINED ||
+                membershipManager.state() == MemberState.FAILED;
+    }
+
+    private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest() {
+        ConsumerGroupHeartbeatRequestData data = new ConsumerGroupHeartbeatRequestData()
+            .setGroupId(membershipManager.groupId())
+            .setMemberEpoch(membershipManager.memberEpoch())
+            .setMemberId(membershipManager.memberId())
+            .setRebalanceTimeoutMs(rebalanceTimeoutMs);
+
+        membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+
+        if (this.subscriptions.hasPatternSubscription()) {
+            // We haven't discsussed how Regex is stored in the consumer. We could do it in the subscriptionState
+            // , in the memberStateManager, or here.
+            // data.setSubscribedTopicRegex(regex)
+        } else {
+            data.setSubscribedTopicNames(new ArrayList<>(this.subscriptions.subscription()));
+        }
+
+        // stubbed
+        data.setServerAssignor("stubbed.assignor");
```

Comment:
Current I'm unable to reference the assignor config in memberStateManager because it is always null.

Reply:
It's null here simply because it wasn't defined when creating the `membershipManager` in the DefaultBackgroundThread.

My understanding is:

- If the client specifies an assignor (client or server), we should make sure to set it in the `membershipManager` when creating it in the DefaultBackgroundThread [here](https://github.com/apache/kafka/blob/d15659a4027698f0f231828315752991575e471c/clients/src/main/java/org/apache/kafka/clients/consumer/internals/DefaultBackgroundThread.java#L193). This will ensure that it is available here to send it on the HeartbeatRequest.

- If the client did not specified any assignor, then it will be OK to have null here, and we don't need to include anything for it in the HeartbeatRequest. We let the group coordinator on the server select the default assignor for the member from the `group.consumer.assignors` config. Correct me here @dajac if I'm missing something.

Reply:
That's right. The (server) assignor should be null be default as defined in the [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-848%3A+The+Next+Generation+of+the+Consumer+Rebalance+Protocol#KIP848:TheNextGenerationoftheConsumerRebalanceProtocol-ConsumerConfigurations). When null, the server uses the first one in the list on the server side.

Reply:
thanks for the clarification.